gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java

```diff
@@ -149,24 +149,34 @@ public void handle(final HttpServerRequest req, final Buffer bodyData, @Nullable
         log.debug("Forwarding request: {} to {}://{} with rule {}", req.uri(), rule.getScheme(), target + targetUri, rule.getRuleIdentifier());
         final String userId = extractUserId(req, log);
         req.pause(); // pause the request to avoid problems with starting another async request (storage)
-        if (userId != null && rule.getProfile() != null && userProfilePath != null) {
-            log.debug("Get profile information for user '{}' to append to headers", userId);
-            String userProfileKey = String.format(userProfilePath, userId);
-            storage.get(userProfileKey, buffer -> {
-                Map<String, String> profileHeaderMap = new HashMap<>();
-                if (buffer != null) {
-                    JsonObject profile = new JsonObject(buffer.toString());
-                    profileHeaderMap = createProfileHeaderValues(profile, log);
-                    log.debug("Got profile information of user '{}'", userId);
-                    log.debug("Going to send parts of the profile in header: {}", profileHeaderMap);
-                } else {
-                    log.debug("No profile information found in local storage for user '{}'", userId);
-                }
-                handleRequest(req, bodyData, targetUri, log, profileHeaderMap, afterHandler);
-            });
-        } else {
-            handleRequest(req, bodyData, targetUri, log, null, afterHandler);
-        }
+
+        maybeAuthenticate(rule).onComplete(event -> {
+            if(event.failed()) {
+                req.resume();
+                log.error("Failed to authenticate request. Cause: {}", event.cause().getMessage());
+                respondError(req, StatusCode.INTERNAL_SERVER_ERROR);
```

Comment:
What happens if access is not granted, we should then return some 401/3, no?

Reply:
I changed the status code to _401 UNAUTHORIZED_